profile/profile.go

```diff
@@ -35,69 +34,148 @@ func InitializeAllProfiles() {
 		ErrorList:   make([]error, 0, 10),
 	}
 
-	defer func() {
-		summaryChan <- summary
-	}()
-
-	err := initEventMap(profile)
+	err := initProcedureList(profile)
 	if err != nil {
 		summary.ErrorList = append(summary.ErrorList, err)
+		summaryChan <- summary
 		return
 	}
-	err = initProcedureList(profile)
+
+	imsi, err := strconv.Atoi(profile.StartImsi)
 	if err != nil {
+		err = fmt.Errorf("invalid imsi value:%v", profile.StartImsi)
 		summary.ErrorList = append(summary.ErrorList, err)
+		summaryChan <- summary
 		return
 	}
+	startImsi := imsi
+	profile.Imsi = imsi
+
+	profile.Log.Infoln("Init profile:", profile.Name,
+		", profile type:", profile.ProfileType)
+
+	if profile.PerUserTimeout == 0 {
+		profile.PerUserTimeout = profctx.PER_USER_TIMEOUT
+	}
 
 	gnb, err := factory.AppConfig.Configuration.GetGNodeB(profile.GnbName)
 	if err != nil {
 		err = fmt.Errorf("Failed to fetch gNB context: %v", err)
 		summary.ErrorList = append(summary.ErrorList, err)
+		summaryChan <- summary
 		return
 	}
 
-	imsi, err := strconv.Atoi(profile.StartImsi)
-	if err != nil {
-		err = fmt.Errorf("invalid imsi value:%v", profile.StartImsi)
-		summary.ErrorList = append(summary.ErrorList, err)
-		return
+	for count := 1; count <= profile.UeCount; count++ {
+		imsiStr := "imsi-" + strconv.Itoa(startImsi)
+		initImsi(profile, gnb, imsiStr)
+		startImsi++
 	}
+}
 
-	profile.Log.Infoln("executing profile:", profile.Name,
-		", profile type:", profile.ProfileType)
+func initImsi(profile *profctx.Profile, gnb *gnbctx.GNodeB, imsiStr string) {
+	readChan := make(chan *common.ProfileMessage)
+	c := simue.InitUE(imsiStr, gnb, profile, readChan)
+	p := profctx.ProfileUeContext{WriteSimChan: c}
+	p.CurrentItr = profile.StartIteration
+	p.ReadChan = readChan
+	trigChan := make(chan *common.ProfileMessage)
+	p.TrigEventsChan = trigChan
+	p.Log = logger.ProfUeCtxLog.WithField(logger.FieldSupi, imsiStr)
+	profile.PSimUe[imsiStr] = &p
+}
 
-	if profile.PerUserTimeout == 0 {
-		profile.PerUserTimeout = profctx.PER_USER_TIMEOUT
-	}
+// option1 : Run default profile start to end..Once done Received
+//    - PROFILE_PASS
+//    - PROFILE_FAIL
+// option2 : Run custom profile start to end thorugh iterations. Once done receive
+//    - PROFILE_PASS
+//    - PROFILE_FAIL
+//    - PROFILE_HOLD optionally, if next iteration is hold
+// option3 : Pulses to come from REST Api
+//    - Start, End, Run-x, Run-y, runx/imsi1, rumy/imsi2)
+//    - Hold the calls
+//    - We should be able to pass events to profile
+
+func ExecuteProfile(profile *profctx.Profile, summaryChan chan common.InterfaceMessage) {
 
+	profile.Log.Infoln("ExecuteProfile started ")
 	var wg sync.WaitGroup
 	var Mu sync.Mutex
-	// Currently executing profile for one IMSI at a time
+
+	summary := &common.SummaryMessage{
+		ProfileType: profile.ProfileType,
+		ProfileName: profile.Name,
+		ErrorList:   make([]error, 0, 10),
+	}
+
+	defer func() {
+		summaryChan <- summary
+	}()
+
+	go func() {
```

Comment:
what does this function do compared the one below it?

Reply:
This function waits for trigger to -> add new call trigger at profile level.

Reply:
These are add on IMSIs. 
we can perhaps do better job, but may be later